Fuse module enhancements

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26457 Fuse module enhancements**

Enhancement to fuse module to support sequentials, fuse list can now be just like the state dict.
Also add support for Conv-Relu and linear-relu fusion

Differential Revision: [D17466382](https://our.internmc.facebook.com/intern/diff/D17466382/)

raghuramank10000 committed:
Fuse module enhancements

Enhancement to fuse module to support sequentials, fuse list can now be just like the state dict.
Also add support for Conv-Relu and linear-relu fusion

Differential Revision: [D17466382](https://our.internmc.facebook.com/intern/diff/D17466382/)

[ghstack-poisoned]

dzhulgakov commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -246,17 +246,16 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                                          stride=stride, padding=padding, dilation=dilation,
-                                         groups=groups, bias=bias, padding_mode=padding_mode)
+                                         groups=groups, bias=bias, padding_mode=padding_mode,
+                                         qconfig = qconfig)
```
nit: formatting

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
+    tokens = submodule_key.split('.')
+    cur_mod = model
+    for s in tokens:
+        if s.isdigit():
```
I just tried and I don't think you need this - getattr seems to work just fine on string keys that are numbers

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
+    cur_mod = model
+    prev_mod = model
+    for s in sub_tokens:
+        if s.isdigit():
```
same

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
+        else:
+            cur_mod = getattr( cur_mod, s)
+
+    if tokens[-1].isdigit():
```
same

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -3,6 +3,15 @@
+    """
+    return torch.nn._intrinsic.LinearReLU(linear, relu)
 
 def fuse_conv_bn(conv, bn):
```
this function (and probably `fuse_conv_bn_eval`) misses setting the training mode. Something like `result.train(original.training)`

raghuramank100 replied:
There is no need to do that here, we just create a sequential here with the original modules, so all attributes are preserved

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
-    new_mod = mod[0]
     if fuser_func is None:
         types = tuple(type(m) for m in mod)
         fuser_func = OP_LIST_TO_FUSER_FUNC.get(types, None)
```
I'd pull this logic into something like 'fuse_known_modules' and pass it as a default to `fuser_func` - it's clear this way and makes it easier to extend functionality

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
+    _set_module(model, modules_to_fuse[0], new_mod)
 
+    for item in modules_to_fuse[1:]:
+        _set_module(model, item, torch.nn.Identity())
```
for cleanness - set the training mode here too

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -111,6 +144,5 @@ def fuse_modules(model, modules_to_fuse):
-    named_module_dict = {name: mod for name, mod in model.named_modules()}
     for module_list in modules_to_fuse:
-        _fuse_modules(model, named_module_dict, module_list)
+        _fuse_modules(model, module_list)
```
can we add a check that if passed in thing is a list of strings just do it? (and update the doc accordingly)

```
if len(modules_to_fuse) > 0 and isinstance(modules_to_fuse, str):
  _fuse_modules(model, modules_to_fuse)
else:
  for module_list ...
```

raghuramank100 replied:
What is passed in is a list of list of strings, like, for example:
```python
        fuse_modules(model, [['conv1', 'relu1'] ,
                             ['features.0.0', 'features.0.1', 'features.0.2'],
                             ['features.1.0', 'features.1.1', 'features.1.2'],
                             ['features.2.0', 'features.2.1', 'features.2.2'],
                             ['classifier.0', 'classifier.1']])
```
So, not sure what you mean here? can you elaborate?

dzhulgakov replied:
I mean also supporting passing a single list of strings (in addition to list of list of strings). It's just convenient to also accept it without extra nesting:

`fuse_modules(model, ['conv1', 'relu1'])`

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
+    for item in modules_to_fuse[1:]:
+        _set_module(model, item, torch.nn.Identity())
 
 def fuse_modules(model, modules_to_fuse):
```
I wonder whether we should do special handling for Sequential. The good thing about sequantials is that we **know** what forward looks like. So we could just iterate through them and try to find a match.

I'm also fine doing it later in a separate PR as it doesn't appear to be a critical feature.

raghuramank100 replied:
This is useful only when we have a sequential module at the leaf of the module tree having fusable modules, so lets do this in a later PR

dzhulgakov reviewed (commented):
Looking good, some more suggestions

jerryzh168 commented on test/test_quantization.py:
```diff
@@ -14,7 +14,7 @@
 from common_quantization import QuantizationTestCase, SingleLayerLinearModel, \
     SkipQuantModel, QuantStubModel, \
-    ModelForFusion, ManualLinearQATModel, ManualConvLinearQATModel, \
+    ModelForFusion, ModelForFusion2, ManualLinearQATModel, ManualConvLinearQATModel, \
```
nit: can you rename ModelForFusion and ModelForFusion2 to more informative names

raghuramank10000 committed:
Update on "Fuse module enhancements"


Enhancement to fuse module to support sequentials, fuse list can now be just like the state dict.
Also add support for Conv-Relu and linear-relu fusion

Differential Revision: [D17466382](https://our.internmc.facebook.com/intern/diff/D17466382/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fuse module enhancements"


Enhancement to fuse module to support sequentials, fuse list can now be just like the state dict.
Also add support for Conv-Relu and linear-relu fusion

Differential Revision: [D17466382](https://our.internmc.facebook.com/intern/diff/D17466382/)

[ghstack-poisoned]

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -1,8 +1,18 @@
+import copy
 
 import torch.nn._intrinsic.modules.fused as torch_fused
+def fuse_conv_relu(conv, relu):
```
you don't need these function - just put torch.nn._intrinsic.ConvReLU2d in the dict below

dzhulgakov commented on torch/quantization/fuse_modules.py:
```diff
@@ -52,44 +62,78 @@ def fuse_conv_bn_relu(conv, bn, relu):
+    for i, item in enumerate(modules_to_fuse):
+        _set_module(model, item, new_mod_list[i])
+
+def fuse_modules(model, modules_to_fuse, inplace=False):
```
you probably wanted to expose fuser_func here too?

dzhulgakov reviewed (approved):
Minor comments - otherwise good to go

raghuramank10000 committed:
Update on "Fuse module enhancements"


Enhancement to fuse module to support sequentials, fuse list can now be just like the state dict.
Also add support for Conv-Relu and linear-relu fusion

Differential Revision: [D17466382](https://our.internmc.facebook.com/intern/diff/D17466382/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Fuse module enhancements"


Enhancement to fuse module to support sequentials, fuse list can now be just like the state dict.
Also add support for Conv-Relu and linear-relu fusion

Differential Revision: [D17466382](https://our.internmc.facebook.com/intern/diff/D17466382/)

[ghstack-poisoned]
